### PR TITLE
refactor: make shell.OpenExternal async

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Offload `OpenExternal` to new thread on Windows.

cc @deepak1556 @zcbenz 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: atom/common/platform_util_win.cc [hunk 1]
```diff
@@ -295,24 +296,29 @@ bool OpenItem(const base::FilePath& full_path) {
...
-    // bug 1136923.
-    return false;
+    // On failure, it may be good to display a message to the user.
+    // https://crbug.com/727913
+    return;
   }
-  return true;
+}
+
+void OpenExternal(const GURL& url) {
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: Use already available `void OpenExternal(const base::string16&, const OpenExternalOptions&, OpenExternalCallback)` method, a new signature is unnecessary.

File: atom/common/platform_util_win.cc [hunk 2]
```diff
@@ -295,24 +296,29 @@ bool OpenItem(const base::FilePath& full_path) {
     return ui::win::OpenFileViaShell(full_path);
 }
 
-bool OpenExternal(const base::string16& url,
-                  const OpenExternalOptions& options) {
+void OpenExternalOnWorkerThread(const base::string16& url,
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: Move this helper function to an anonymous namespace

**Comment by deepak1556** [comment]
Do we want to support both `OpenExternalSync` and `OpenExternal` ? Its not possible to support sync version on windows with this change.

**Comment by zcbenz** [comment]
We can either deprecate the sync API or just leave a fake sync API. I'm good with either, the async API was fake anyway.

File: atom/browser/atom_browser_client.cc [hunk 3]
```diff
@@ -791,15 +791,16 @@ void AtomBrowserClient::RenderProcessExited(
   }
 }
 
+void OnOpenExternalFinished(const std::string& msg) {
+  LOG(INFO) << msg;
```
Comment (nornagon) [inline atom/browser/atom_browser_client.cc]:
```suggestion
```

File: atom/browser/atom_browser_client.cc [hunk 4]
```diff
@@ -791,15 +791,16 @@ void AtomBrowserClient::RenderProcessExited(
...
-#if defined(OS_WIN)
-        base::UTF8ToUTF16(escaped_url.spec()),
-#else
-        escaped_url,
-#endif
-        platform_util::OpenExternalOptions());
+  if (allowed) {
+    platform_util::OpenExternal(escaped_url,
+                                platform_util::OpenExternalOptions(),
+                                base::BindOnce(&OnOpenExternalFinished));
```
Comment (deepak1556) [inline atom/browser/atom_browser_client.cc]: `base::NullCallback` maybe ?

File: atom/browser/atom_browser_client.cc (hunk 3 above)
Comment (codebytere) [reply atom/browser/atom_browser_client.cc]: should it just be empty then do you think?

File: atom/common/platform_util_win.cc [hunk 5]
```diff
@@ -301,31 +320,16 @@ bool OpenItem(const base::FilePath& full_path) {
...
-}
-
-void OpenExternal(const base::string16& url,
+void OpenExternal(const GURL& url,
                   const OpenExternalOptions& options,
                   OpenExternalCallback callback) {
   // TODO(gabriel): Implement async open if callback is specified
-  std::move(callback).Run(OpenExternal(url, options) ? "" : "Failed to open");
+  std::move(callback).Run(
+      base::CreateCOMSTATaskRunnerWithTraits(
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: `callback` should be executed after the `OpenExternalOnWorkerThread` task is complete. Also use the return values from that task to depict errors in the callback accordingly.

File: atom/browser/atom_browser_client.cc (hunk 3 above)
Comment (nornagon) [reply atom/browser/atom_browser_client.cc]: if it does nothing, why have it?

Comment (deepak1556) [reply atom/browser/atom_browser_client.cc]: I think it was intended to adjust to the openExternal api that needed a callback as argument. @codebytere you can use `base::NullCallback()` in such cases.

File: atom/browser/atom_browser_client.cc (hunk 4 above)
Comment (codebytere) [reply atom/browser/atom_browser_client.cc]: til that's a thing! ty :D

File: atom/common/platform_util_win.cc (hunk 5 above)
Comment (codebytere) [reply atom/common/platform_util_win.cc]: done!

File: atom/browser/atom_browser_client.cc (hunk 3 above)
Comment (codebytere) [reply atom/browser/atom_browser_client.cc]: yep! thanks :D

File: atom/common/platform_util_win.cc [hunk 6]
```diff
@@ -230,6 +230,25 @@ HRESULT DeleteFileProgressSink::ResumeTimer() {
...
+  // parameters unexpected by the external program. This url should already
+  // have been escaped.
+  std::string escaped_url = url.spec();
+  auto working_dir = options.working_dir.value();
+
+  if (reinterpret_cast<ULONG_PTR>(
+          ShellExecuteW(nullptr, "open", escaped_url.c_str(), nullptr,
+                        working_dir.empty() ? nullptr : working_dir.c_str(),
+                        SW_SHOWNORMAL)) <= 32) {
+    std::move(callback).Run("Failed to open");
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: `callback` should be invoked on the main thread, this is a different sequence.

File: atom/common/platform_util_win.cc [hunk 7]
```diff
@@ -301,31 +320,14 @@ bool OpenItem(const base::FilePath& full_path) {
...
-
-void OpenExternal(const base::string16& url,
+void OpenExternal(const GURL& url,
                   const OpenExternalOptions& options,
                   OpenExternalCallback callback) {
-  // TODO(gabriel): Implement async open if callback is specified
-  std::move(callback).Run(OpenExternal(url, options) ? "" : "Failed to open");
+  base::CreateCOMSTATaskRunnerWithTraits(
+      {base::MayBlock(), base::TaskPriority::USER_BLOCKING})
+      ->PostTaskReplyWithResult(
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: This should be

```
base::CreateCOMSTATaskRunnerWithTraits(
      {base::MayBlock(), base::TaskPriority::USER_BLOCKING})
      ->PostTaskAndReplyWithResult(FROM_HERE, base::BindOnce(&OpenExternalOnWorkerThread, url, options),
                                   std::move(callback)));
```

Also make the signature of `OpenExternalOnWorkerThread` to be `std::string OpenExternalOnWorkerThread(...)`

File: atom/common/platform_util_win.cc [hunk 8]
```diff
@@ -301,31 +319,15 @@ bool OpenItem(const base::FilePath& full_path) {
...
                   const OpenExternalOptions& options,
                   OpenExternalCallback callback) {
-  // TODO(gabriel): Implement async open if callback is specified
-  std::move(callback).Run(OpenExternal(url, options) ? "" : "Failed to open");
+  std::string msg = base::CreateCOMSTATaskRunnerWithTraits(
+      {base::MayBlock(), base::TaskPriority::USER_BLOCKING})
+      ->PostTaskAndReplyWithResult(FROM_HERE, base::BindOnce(&OpenExternalOnWorkerThread, url, options),
+                                   std::move(callback)));
+
+  std::move(callback).Run(msg);
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: This line is not required. `callback` is now task that will be called with the result from `OpenExternalOnWorkerThread` in the previous call.

File: atom/common/platform_util_win.cc [hunk 9]
```diff
@@ -301,31 +319,15 @@ bool OpenItem(const base::FilePath& full_path) {
...
-  return true;
-}
-
-void OpenExternal(const base::string16& url,
+void OpenExternal(const GURL& url,
                   const OpenExternalOptions& options,
                   OpenExternalCallback callback) {
-  // TODO(gabriel): Implement async open if callback is specified
-  std::move(callback).Run(OpenExternal(url, options) ? "" : "Failed to open");
+  std::string msg = base::CreateCOMSTATaskRunnerWithTraits(
```
Comment (deepak1556) [inline atom/common/platform_util_win.cc]: There are no synchronous responses here.

**Comment by codebytere** [comment]
@deepak1556 this should finally be ready for final review!

**Pull request merged**